portfolio/src/main/java/com/google/sps/servlets/BookServlet.java

```diff
@@ -23,48 +21,14 @@
   private Map<Integer, Book> books;
 
   @Override
-  public void init() throws ServletException {
-    books = readBooks();
-  }
-
-  private HashMap<Integer, Book> readBooks() throws ServletException {
-    HashMap<Integer, Book> allBooks = new HashMap<Integer, Book>();
-    String path = "/WEB-INF/"
-        + "20_books.csv";
-    try (Scanner scanner =
-             new Scanner(getServletContext().getResourceAsStream(path)).useDelimiter("\\Z")) {
-      String content = scanner.next().replaceAll("[\\r\\n]+", "");
-      String[] lines = content.split("NEXTBOOK"); // lines[i] represents one row of the file
-
-      String current_title = "";
-      Book.Builder current_builder = Book.builder().title("null");
-      int currentId = 0;
+  public void init() {
 
-      for (int i = 1; i < lines.length; i++) {
-        String[] cells = lines[i].split(",");
-        String title = cells[0];
-        String genre = cells[8];
-        String review = cells[13];
-        if (current_title.equals(title)) {
-          // add review:
-          current_builder.addReview(review);
-        } else {
-          // close old book:
-          if (i != 1) {
-            Book book = current_builder.build();
-            allBooks.put(currentId++, book);
-          }
-          // start building new book
-          current_builder = Book.builder().title(title).genre(genre).addReview(review);
-          current_title = title;
-        }
-      }
-      Book book = current_builder.build();
-      allBooks.put(currentId, book);
-    } catch (Exception e) {
-      throw new ServletException("Error reading CSV file", e);
+    try{
+        BookReader reader = new BookReader(getServletContext().getResourceAsStream("/WEB-INF/20_books.csv"));
```

Comment:
Passing InputStream as constructor arg is discouraged, because it makes it unclear who should close the stream, how will the exception be handled, and if the mafebookList method is not invoked, it will never be closed.

I think. you could get the real path given the virtual path using `getServletContext().getRealPath("/WEB-INF/20_books.csv")`: https://docs.oracle.com/javaee/7/api/javax/servlet/ServletContext.html#getRealPath-java.lang.String-

If this doesn't work, you can pass the relative path to the BookReader class, and use `BookReader.class.getResourceAsStream(path)` from within BookReader class : https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getResourceAsStream(java.lang.String)